File: Problems/BestTimeToBuySellStock/best_time_to_buy_sell_stock.py
"""
Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

Example 1:

Input: nums = [1,2,3,1]
Output: true
"""

from typing import List
import json


def buy_sell_stock(prices:List[int])->int:
    max_profit = 0

    buy = 0
    sell = 1
    
    while sell < len(prices):
        # calculate the difference between the sell and buy point.
        delta = prices[sell] - prices[buy]

        # if it's positive
        if delta > 0:
            # that means we made a profit.  
            max_profit = max(max_profit, delta)
        else:
            # a loss...we can discard negative deltas. move buy up to sell.
            buy = sell


        # in either case, we'll move our sell pointer to the right
        sell += 1
    
    return max_profit

if __name__ == "__main__":
    with open("test_cases.json", mode='r', encoding="utf-8-sig") as file:
        text = file.read() 
        test_cases = json.loads(text)
        for test in test_cases['tests']:
            input = test['input']
            expected = test['expected']
            result = buy_sell_stock(prices=input)
            passed = expected == result
            print(f"Input:{input}\nResult: {result}\nExpected: {expected}\nPassed: {passed}\n\n")

File: Problems/BestTimeToBuySellStock/best_time_to_buy_sell_stock.go
/* 
Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

Example 1:

Input: nums = [1,2,3,1]
Output: true
*/
package main

import (
	"fmt"
	"encoding/json"
    "io/ioutil"
	"log"
)

type TestCase struct {
	Input []int
	Expected int
	Comment string
}
type TestCases struct {
	Tests []TestCase
}

// helper function to find max value of two integers.
// (built in math.Max function only works on floats)
func max(val1 int, val2 int) int {
	if val1 < val2 {
		return val2
	}

	return val1
}

func buySellStock(prices []int) int {
	var maxProfit = 0
	var buy = 0
	var sell = 1

	for sell < len(prices) {
		// calculate the difference at the sell and buy points.
		var delta = prices[sell] - prices[buy]

		// if we see a profit over that interval
		if delta > 0 {
			// sell price was higher than buy price.
			// calculate the max so far.
			maxProfit = max(maxProfit, delta)
		} else {
			// sell price was lower than buy. we saw a loss, so, we can move our buy up to sell. 
			// this is minimizing the buy price.
			buy = sell
		}

		// always move the sell pointer to the right
		sell += 1
	}

	return maxProfit
}

func main() {
	content, err := ioutil.ReadFile("./test_cases.json")
    if err != nil {
        log.Fatal("Error when opening file: ", err)
    }

	var testCases TestCases
    err = json.Unmarshal(content, &testCases)
    if err != nil {
        log.Fatal("Error during Unmarshal(): ", err)
		
    }

	for _, test := range testCases.Tests {
		var input = test.Input
		var expected = test.Expected
		var result = buySellStock(input)
		
		// a bit more work here for Go
		var passed = result == expected
		fmt.Printf("Input: %v\nResult: %d\nExpected: %d\nPassed: %v\n\n", input, result, expected, passed)
	}

}